runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MetricResultsProtos.java

```diff
@@ -0,0 +1,235 @@
+      MetricKey metricKey,
+      T value,
+      BiConsumer<SimpleMonitoringInfoBuilder, T> set,
+      BiConsumer<BeamFnApi.MetricResults.Builder, MonitoringInfo> add) {
+    if (value != null) {
+      SimpleMonitoringInfoBuilder partial =
+          new SimpleMonitoringInfoBuilder().handleMetricKey(metricKey);
+      ;
+      set.accept(partial, value);
+      MonitoringInfo monitoringInfo = partial.build();
+      if (monitoringInfo != null) {
+        add.accept(builder, monitoringInfo);
+      }
+    }
+  }
+
+  // `fromProto` helpers
+
+  /**
+   * Helper for converting {@link BeamFnApi.MetricResults} to {@link MetricResults}.
+   *
+   * <p>The former separates "attempted" and "committed" metrics at the highest level, while the
+   * latter splits on metric-type (counter, distribution, or gauge) first, so converting basically
+   * amounts pivoting those between those two.
+   */
+  private static class MetricResultsBuilder {
+    private final Map<MetricKey, MetricResultBuilder<Long>> counters = new ConcurrentHashMap<>();
+    private final Map<MetricKey, MetricResultBuilder<DistributionResult>> distributions =
+        new ConcurrentHashMap<>();
+    private final Map<MetricKey, MetricResultBuilder<GaugeResult>> gauges =
+        new ConcurrentHashMap<>();
+
+    /**
+     * Populate metric-type-specific maps with {@link MonitoringInfo}s from a {@link
+     * BeamFnApi.MetricResults}.
+     */
+    public MetricResultsBuilder(BeamFnApi.MetricResults metrics) {
+      add(metrics.getAttemptedList(), false);
+      add(metrics.getCommittedList(), true);
+    }
+
+    private void add(Iterable<MonitoringInfo> monitoringInfos, Boolean committed) {
+      for (MonitoringInfo monitoringInfo : monitoringInfos) {
+        add(monitoringInfo, committed);
+      }
+    }
+
+    public MetricResults build() {
+      return new DefaultMetricResults(build(counters), build(distributions), build(gauges));
+    }
+
+    /** Helper for turning a map of result-builders into a sequence of results. */
+    private static <T> List<MetricResult<T>> build(Map<MetricKey, MetricResultBuilder<T>> map) {
+      return map.values().stream().map(MetricResultBuilder::build).collect(toList());
+    }
+
+    private void add(MonitoringInfo monitoringInfo, Boolean committed) {
+      add(
+          keyFromMonitoringInfo(monitoringInfo),
+          monitoringInfo.getType(),
+          monitoringInfo.getTimestamp(),
+          monitoringInfo.getMetric(),
+          committed);
+    }
+
+    private void add(
+        MetricKey metricKey, String type, Timestamp timestamp, Metric metric, Boolean committed) {
+      processMetric(
+          metric,
+          type,
+          timestamp,
+          counter -> add(metricKey, counter, committed, counters),
+          distribution -> add(metricKey, distribution, committed, distributions),
+          gauge -> add(metricKey, gauge, committed, gauges));
+    }
+
+    private <T> void add(
```

Comment:
With this many params, its not clear what's geting added, please change the name